lib/widgets/compose_box.dart

```diff
@@ -210,85 +213,212 @@ class _StreamContentInputState extends State<_StreamContentInput> {
+  _File({required this.content, required this.length, required this.filename});
 
-  final ContentTextEditingController contentController;
-  final FocusNode contentFocusNode;
+  final Stream<List<int>> content;
+  final int length;
+  final String filename;
+}
 
-  _handlePress(BuildContext context) async {
-    FilePickerResult? result;
+Future<void> _uploadFiles({
+  required BuildContext context,
+  required ContentTextEditingController contentController,
+  required FocusNode contentFocusNode,
+  required Iterable<_File> files,
+}) async {
+  assert(context.mounted);
+  final store = PerAccountStoreWidget.of(context);
+
+  final List<_File> tooLargeFiles = [];
+  final List<_File> rightSizeFiles = [];
+  for (final file in files) {
+    if ((file.length / (1 << 20)) > store.maxFileUploadSizeMib) {
+      tooLargeFiles.add(file);
+    } else {
+      rightSizeFiles.add(file);
+    }
+  }
+
+  if (tooLargeFiles.isNotEmpty) {
+    final listMessage = tooLargeFiles
+      .map((file) => '${file.filename}: ${(file.length / (1 << 20)).toStringAsFixed(1)} MiB')
+      .join('\n');
+    showErrorDialog( // TODO(i18n)
+      context: context,
+      title: 'File(s) too large',
+      message:
+        '${tooLargeFiles.length} file(s) are larger than the server\'s limit of ${store.maxFileUploadSizeMib} MiB and will not be uploaded:\n\n$listMessage');
+  }
+
+  final List<(int, _File)> uploadsInProgress = [];
+  for (final file in rightSizeFiles) {
+    final tag = contentController.registerUploadStart(file.filename);
+    uploadsInProgress.add((tag, file));
+  }
+  if (!contentFocusNode.hasFocus) {
+    contentFocusNode.requestFocus();
+  }
+
+  for (final (tag, file) in uploadsInProgress) {
+    final _File(:content, :length, :filename) = file;
+    Uri? url;
     try {
-      result = await FilePicker.platform.pickFiles(allowMultiple: true, withReadStream: true);
+      final result = await uploadFile(store.connection,
+        content: content, length: length, filename: filename);
+      url = Uri.parse(result.uri);
     } catch (e) {
-      // TODO(i18n)
-      showErrorDialog(context: context, title: 'Error', message: e.toString());
-      return;
+      if (!context.mounted) return;
+      // TODO(#37): Specifically handle `413 Payload Too Large`
+      // TODO(#37): On API errors, quote `msg` from server, with "The server said:"
+      showErrorDialog(context: context,
+        title: 'Failed to upload file: $filename', message: e.toString());
+    } finally {
+      contentController.registerUploadEnd(tag, url);
     }
-    if (result == null) {
-      return; // User cancelled; do nothing
+  }
+}
+
+abstract class _AttachUploadsButton extends StatelessWidget {
+  const _AttachUploadsButton({required this.contentController, required this.contentFocusNode});
+
+  final ContentTextEditingController contentController;
+  final FocusNode contentFocusNode;
+
+  IconData get icon;
+  Future<Iterable<_File>?> getFiles(BuildContext context);
```

Comment:
It'd be good to have a bit of dartdoc, particularly to say what `null` means.

… Hmm, actually, perhaps better: make it non-nullable, and just return an empty list in that case. Then at the call site, skip doing anything if the list is empty (rather than if null).

That will even behave differently from this version if there's some way that `pickFiles` or whatever can return an empty list: it'll do nothing in that case, whereas this version would go and focus the content input. And I think doing nothing is appropriate — if it does return an empty list, that probably means the user has effectively tried to abort the operation.